pkg/kubelet/cm/topologymanager/topology_manager.go

```diff
@@ -164,16 +164,35 @@ func (m *manager) calculateAffinity(pod v1.Pod, container v1.Container) Topology
 		// Get the TopologyHints from a provider.
 		hints := provider.GetTopologyHints(pod, container)
 
-		// If hints is nil, overwrite 'hints' with a preferred any-socket affinity.
+		// If hints is nil, insert a single, preferred any-socket hint into allProviderHints.
 		if hints == nil || len(hints) == 0 {
-			klog.Infof("[topologymanager] Hint Provider has no preference for socket affinity")
+			klog.Infof("[topologymanager] Hint Provider has no preference for socket affinity with any resource")
 			affinity, _ := socketmask.NewSocketMask()
 			affinity.Fill()
-			hints = []TopologyHint{{affinity, true}}
+			allProviderHints = append(allProviderHints, []TopologyHint{{affinity, true}})
+			continue
 		}
 
-		// Accumulate the sorted hints into a [][]TopologyHint slice
-		allProviderHints = append(allProviderHints, hints)
+		// Otherwise, accumulate the hints for each resource type into allProviderHints.
+		for resource := range hints {
+			if hints[resource] == nil {
+				klog.Infof("[topologymanager] Hint Provider has no preference for socket affinity with resource '%s'", resource)
+				affinity, _ := socketmask.NewSocketMask()
+				affinity.Fill()
+				allProviderHints = append(allProviderHints, []TopologyHint{{affinity, true}})
+				continue
+			}
+
+			if len(hints[resource]) == 0 {
+				klog.Infof("[topologymanager] Hint Provider has no possible socket affinities for resource '%s'", resource)
+				affinity, _ := socketmask.NewSocketMask()
+				affinity.Fill()
+				allProviderHints = append(allProviderHints, []TopologyHint{{affinity, false}})
+				continue
+			}
+
+			allProviderHints = append(allProviderHints, hints[resource])
```

Comment:
So you going over all the resources just to change them from a map[string][]TopologyHint to  [][]TopologyHint. Why not just keeping it as a map[string][]TopologyHint. In my merge abstraction I was expecting the permutation to be map[string]TopologyHint. I was thinking that it might be useful in the future that we will know what is the resource this hint is belong too. One case for that is GPU direct. I will be able to write a policy that align GPU PCI  address with NIC PCI address which are have on the same NUMA with the cpus.

Reply:
I am fine keeping it as a map if that makes your `Merge()` abstraction better. I was just tring to change the code from its current state as little as possible.

Reply:
Is this due time constraints for the release? not sure when kubernetes 1.16 is release, and I doubt the merge abstraction will be merged into this release. (there are other patches that should be merged first). If that the case we can keep the code as is, but we will need to revisit this later.

Reply:
It's more due to ease of reviewability. the smaller the change, the easier it is to review.

Reply:
@moshe010 You could introduce the change in Topology Manager as part of your PR and explain the reasoning

Reply:
yes, sure I will do the changes in my PR

Reply:
on second thought, for the current policies (best effort and strict) I don't need merge  signature to be  map only for more advance policies in the future.  Because we are in time constraints for 1.16 I will make the merge signature to be slice and we will revisit it late.